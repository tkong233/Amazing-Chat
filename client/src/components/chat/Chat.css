.chat-container {
  position: relative;
  width: 100%;
  height: 80vh;
}

.chat-input-bar {
  /* position: absolute; */
  /* bottom: 0; */
  width: 80%;
  margin-top: 40px;
  margin-left: 10%;
  margin-right: 5%;
  margin-bottom: 2%;
}
