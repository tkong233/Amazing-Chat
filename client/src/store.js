import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from 'redux-persist'
import rootReducer from "./reducers";
import storage from 'redux-persist/lib/storage' // defaults to localStorage for web

const initialState = {};

const persistConfig = {
  key: 'root',
  storage,
}

const middleware = [thunk];

const persistedReducer = persistReducer(persistConfig, rootReducer, initialState);

const store = createStore(

  persistedReducer,
  compose(
    applyMiddleware(...middleware),
    (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()) ||
      compose
  )
);

const persistor = persistStore(store);

export {
  store,
  persistor,
  middleware
};



