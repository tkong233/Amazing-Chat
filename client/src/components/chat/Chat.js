import React, { useState } from "react";
import { connect } from "react-redux";
import Button from "@material-ui/core/Button";
import VideoCallIcon from "@material-ui/icons/VideoCall";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogTitle from "@material-ui/core/DialogTitle";
import {
  sendMessage,
  uploadChatFiles,
  initiateVideoCall,
  pickUpVideoCall,
  hangUpVideoCall,
  waitForVideoCallDecision,
  setCalleeOnline,
  stopWaiting,
} from "../../actions/chatActions";
import Dropzone from "react-dropzone";
import Messages from "./Messages";
import VideoCall from "./VideoCall";
import "./Chat.css";

const Chat = (props) => {
  const {
    messages,
    socket,
    pairId,
    sender,
    receiver,
    receiverName,
  } = props.chat;
  const { name, email } = props.user;

  const [message, setMessage] = useState("");

  const sendMessage = (e) => {
    e.preventDefault();

    if (socket && message && pairId && receiver) {
      // console.log('sending message: ' + message);
      const data = {
        pairId,
        from: sender,
        to: receiver,
        message,
        type: "text",
      };
      props.sendMessage(data, socket);
      setMessage("");
    }
  };

  const [errorUpload, setErrorUpload] = useState("");
  const onDrop = async (files) => {
    setErrorUpload("");
    const file_size = files[0].size; //byte
    if (file_size <= 11000000) {
      const formData = new FormData();
      const { pairId, socket, sender, receiver } = props.chat;
      formData.append("NonTextfile", files[0]);
      formData.append("pairId", pairId);
      formData.append("from", sender);
      formData.append("to", receiver);
      formData.append("type", "ImageOrVideoOrAudio");
      await props.uploadChatFiles(formData, socket);
    } else {
      setErrorUpload("File too large, maximum size 11M");
    }
  };

  const initiateVideoCall = () => {
    // console.log("initiating video call");
    props.initiateVideoCall(pairId, sender, receiver, socket);
    props.waitForVideoCallDecision(socket);
  };

  const handleAcceptVideoCall = () => {
    props.pickUpVideoCall(socket, pairId, sender, receiver);
  };

  const handleRejectVideoCall = () => {
    props.hangUpVideoCall(socket, pairId, sender, receiver);
  };

  const handleCalleeOffline = () => {
    props.setCalleeOnline();
    props.stopWaiting();
  };

  const { ringing, calling, waiting, calleeOnline } = props.chat;
  console.log(
    "ringing",
    ringing,
    "calling",
    calling,
    "waiting",
    waiting,
    "calleeOnline",
    calleeOnline
  );
  console.log(props.chat);
  if (socket && !calling) {
    return (
      // Video Call Dialog
      <div className="chat-container" data-test="ChatComponent">
        {/* Receive Call Window */}
        <Dialog
          open={ringing}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{`${receiverName} invited you to a video call`}</DialogTitle>
          <DialogActions>
            <Button onClick={handleRejectVideoCall} color="primary">
              Reject
            </Button>
            <Button onClick={handleAcceptVideoCall} color="primary" autoFocus>
              Accept
            </Button>
          </DialogActions>
        </Dialog>

        {/* Waiting for Response Window */}
        <Dialog
          open={waiting && calleeOnline}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{`Waiting for ${receiverName} to respond...`}</DialogTitle>
          <DialogActions>
            <Button onClick={handleRejectVideoCall} color="primary">
              Hang Up
            </Button>
          </DialogActions>
        </Dialog>

        {/* On Callee Offline */}
        <Dialog
          open={!calleeOnline}
          onClose={handleCalleeOffline}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{`${receiverName} is not online`}</DialogTitle>
          <DialogActions>
            <Button onClick={handleCalleeOffline} color="primary">
              OK
            </Button>
          </DialogActions>
        </Dialog>

        {/* Messages */}
        <Messages />

        {/* Chat Input */}
        <div className="chat-input-bar" style={{}}>
          <div>
            <input
              id="message-input"
              type="text"
              value={message}
              onChange={(e) => setMessage(e.target.value)}
              onKeyPress={(e) => (e.key === "Enter" ? sendMessage(e) : null)}
            />
            <label htmlFor="message-input">type a message ‚èé</label>
          </div>
          <div style={{display: "flex"}}>
            <Dropzone onDrop={onDrop}>
              {({ getRootProps, getInputProps }) => (
                <section>
                  <div {...getRootProps()}>
                    <input {...getInputProps()} />
                    <Button
                      style={{ color: "white", backgroundColor: "#455a64", marginTop: "7px" }}
                      variant="contained"
                      component="span"
                      size="small"
                    >
                      Send Files (Image, Audio, Video)
                    </Button>
                  </div>
                </section>
              )}
            </Dropzone>
            {/* <VideoCallIcon
            style={{ fontSize: 50, marginTop: "3px" }}
            onClick={initiateVideoCall}
          /> */}
            <div>
              <Button
                // variant="contained"
                // component="span"
                // size="large"
                onClick={initiateVideoCall}
                style={{ color: "CornflowerBlue", backgroundColor: "transparent", marginTop: "2px" }}
              >
                <i class="fas fa-video fa-2x"></i>
              </Button>
            </div>
            <div style={{marginTop: "10px"}} className= "blue-grey-text text-darken-2"> {errorUpload} </div>
          </div>
        </div>
      </div>
    );
  } else if (socket && calling) {
    return <VideoCall />;
  } else {
    return null;
  }
};

const mapStateToProps = (state) => ({
  user: state.auth.user,
  chat: state.chat,
});

export default connect(mapStateToProps, {
  sendMessage,
  uploadChatFiles,
  initiateVideoCall,
  pickUpVideoCall,
  hangUpVideoCall,
  waitForVideoCallDecision,
  setCalleeOnline,
  stopWaiting,
})(Chat);
