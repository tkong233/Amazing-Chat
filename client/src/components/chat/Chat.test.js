import React from 'react';
import Chat from './Chat';
import {shallow} from 'enzyme';
import checkPropTypes from 'check-prop-types';
import Enzyme from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16'; 
import { applyMiddleware, createStore } from 'redux';
import {middleware} from '../../store';
import rootReducer from '../../reducers/index'

Enzyme.configure({
    adapter: new EnzymeAdapter(),
});

const setUp = (initialState={}) =>{

    const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
    const store = createStoreWithMiddleware(rootReducer, initialState);
    const wrapper = shallow(<Chat store={store} />).dive();
    // console.log(wrapper.debug());
    return wrapper;
}

describe('Chat Component', ()=>{
    describe('Checking PropTypes', ()=>{
        it('Should not throw a warning', ()=>{
            const expectedProps = {
                chat: {},
                user: {},
            }
            // eslint-disable-next-line react/forbid-foreign-prop-types
            const propsErr = checkPropTypes(Chat.propTypes, expectedProps, 'props', Chat.name);
            expect(propsErr).toBeUndefined();
        })
    });
    describe('Renders', ()=>{
        let wrapper;
        beforeEach(()=>{
            const initialState = {
                auth:{user: {
                    name: 'testuser',
                    email: 'testuser@test.com',
                    profile_picture: 'testpicture'
                }},
                chat:{
                    messages:[],
                    socket: ''
                }
            };
            wrapper = setUp(initialState);
        });

        // it('Should render without errors', ()=>{
        //     const container = wrapper.find(`[data-test='ChatComponent']`);
        //     expect(container.length).toBe(1);
        // });

        it('Snapshot testing', ()=>{
            expect(wrapper).toMatchSnapshot();
        });


    });
});