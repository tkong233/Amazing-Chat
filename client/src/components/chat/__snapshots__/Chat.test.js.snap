// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Component Renders Snapshot testing 1`] = `ShallowWrapper {}`;
